### PR TITLE
feat(inputs): New lustre2 plugins

### PR DESCRIPTION
**Pull request by Ephemeraler**
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format]

Added an input plugin to gather lustre data with command `lctl get_param`, compared to existing lustre2 plugin, I use command to gather metrics, which is lustre official suggestion.

### CODE REVIEW

**Comment by powersj** [comment]
Hi @Ephemeraler,

Thanks for the PR. Looks like there are some lint and CI issues to resolve.

> which is lustre official suggestion.

Can you please link to where this suggestion is made?

Couple concerns:

1) What does this collect that the current plugin does not? Help me understand the value in having two different plugins for lustre2. Should we extend the existing one instead?
2) We strongly discourage adding lots of boolean config options. This is annoying for users and leads to a never-ending list of additional config options. We prefer to have a single string array that include a list of the possible metrics to collect. (e.g. `collect = ["client", "stats"]`
3) The comments above each function stating the `@param` do not provide any important details and are unnecessary, please remove them.
4) Please look at how we have tests in other plugins using a directory full of test data with expected output. It helps reduce the amount of output that needs to be in the go files.

**Review by powersj — CHANGES_REQUESTED** [review]
Please see my comment

**Comment by Ephemeraler** [comment]
@powersj Thanks, I will accept your suggestions.

**Comment by Ephemeraler** [comment]
@powersj 
1. This [https://wiki.lustre.org/Lustre_Monitoring_and_Statistics_Guide](url) - section "Reading /proc vs lctl" explains why 'lctl get_param' is the standard and recommended way to monitor and statistics. Plugins for lustre2 use /proc file. This is the main reason I want to add plugins `lustre2_lctl`.
2. I will discourage adding lots of boolean config options and use  single string array.
3. I will remove the comments above each function stating the @param.
4. I will refer to other testing file

**Comment by Ephemeraler** [comment]
@powersj Excuse, I have three item unpassed
1) test-go-linux and test-go-linux-386, it stops at `GOARCH=amd64 ./gotestsum -- ${RACE} -short -cover -coverprofile=coverage.out ./...`. But , I have executed this command in my development environment. I found that the result returned by my fakeExecCommand  includes `GOCOVERDIR not set, no coverage data emitted`, so it merge with my expected result, so that my function couldnt parse it.

2) test-go-windows, I think this problem may not have been caused by me.

Thanks.

**Comment by powersj** [comment]
> test-go-windows, I think this problem may not have been caused by me.

This is fixed in our upstream master. If you rebase on master you will get the fix

> I found that the result returned by my fakeExecCommand includes GOCOVERDIR not set, no coverage data emitted, so it merge with my expected result, so that my function couldnt parse it.

Can you check for that output and strip it before comparing it?

File: plugins/inputs/lustre2_lctl/lustre2_lctl.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+package lustre2_lctl
+
+import (
+	_ "embed"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const namespace = "lustre2"
```
Comment (powersj) [inline plugins/inputs/lustre2_lctl/lustre2_lctl.go:13]: I'd rather us call this plugin_name

File: plugins/inputs/lustre2_lctl/lustre2_lctl.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	OstCollect    []string        `toml:"ost_collect"`
+	MdtCollect    []string        `toml:"mdt_collect"`
+	ClientCollect []string        `toml:"client_collect"`
+	Log           telegraf.Logger `toml:"-"`
+}
+
+func (*Lustre2Lctl) SampleConfig() string {
+	return sampleConfig
+}
+
```
Comment (powersj) [inline plugins/inputs/lustre2_lctl/lustre2_lctl.go:30]: I would like to see an `Init` function in here that ensures that the user has access to the `lctl` command before we start up and fail right after we try to gather.

File: plugins/inputs/lustre2_lctl/README.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
+# Lustre Input Plugin
```
Comment (powersj) [inline plugins/inputs/lustre2_lctl/README.md:1]: Let's distinguish this from the other lustre plugin

```suggestion
# Lustre lctl Input Plugin
```

**Review by powersj — COMMENTED** [review]
Hi,

Thanks for the updates. I have done a high-level review but I realized there was one question from last time that was not answered:

What metrics does this capture that the existing plugin does not already provide?

I am very weary of parsing command output and prefer reading the files in /proc. So before we take on a second lustre2 plugin I want to understand the value in maintaining it as well.

Thanks

**Comment by Ephemeraler** [comment]
> Hi,
> 
> Thanks for the updates. I have done a high-level review but I realized there was one question from last time that was not answered:
> 
> What metrics does this capture that the existing plugin does not already provide?
> 
> I am very weary of parsing command output and prefer reading the files in /proc. So before we take on a second lustre2 plugin I want to understand the value in maintaining it as well.
> 
> Thanks

First, This [https://wiki.lustre.org/Lustre_Monitoring_and_Statistics_Guide](https://github.com/influxdata/telegraf/pull/url) - section "Reading /proc vs lctl" explains why 'lctl get_param' is the standard and recommended way to monitor and statistics. Plugins for lustre2 use /proc file. This is the main reason I want to add plugins lustre2_lctl.

Second, metrics like lustre2_client_mdc_volume_active, lustre2_client_ost_volume_active and so on, you couldn't read it from /proc.

File: plugins/inputs/lustre2_lctl/README.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+```
+
+## Metrics
+
+### OST
+
+* tags
+  * volume (the name of volume)
+  * jobid
+  * unit
```
[the remark this answers is not in the document]
Comment (Ephemeraler) [reply plugins/inputs/lustre2_lctl/README.md:59]: jobid and unit is gathered from job_stats or stats, the jobid is job ID, the uint is what statistics are the operations based on. Uints for the same operation is different in different lustre version. for example,

[Lustrev2.15]
destroy                   5473 samples [usecs] 23 227364 19113615 505308014521

[Lustrev2.12]
destroy                   44292747 samples [reqs]

File: plugins/inputs/lustre2_lctl/client.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+				}, map[string]string{
+					"volume": volume,
+				})
+			}
+		}
+	}
+}
+
+func gatherClient(collect []string, namespace string, acc telegraf.Accumulator) {
+	measurement := namespace + "_client"
```
[the remark this answers is not in the document]
Comment (Ephemeraler) [reply plugins/inputs/lustre2_lctl/client.go:58]: namespace is a const string, it was lustre2, but I will change it to lustre2_lctl(plugin_name)

File: plugins/inputs/lustre2_lctl/README.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  * volume
+* fields
+  * osc_volume_active (uint)
+  * mdc_volume_active (uint)
+
+## Troubleshooting
+
+Check for the default or custom procfiles in the proc filesystem, and reference
+the [Lustre Monitoring and Statistics Guide][guide].  This plugin does not
+report all information from these files, only a limited set of items
```
[the remark this answers is not in the document]
Comment (Ephemeraler) [reply plugins/inputs/lustre2_lctl/README.md:99]: Yes, I only miss /proc/fs/lustre/mdt/*/exports/*/stats in Lustre2 plugin. The rest are the same, I replace some words with *.  And compared to lustre2, I gather volume active in lustre client. For /proc/fs/lustre/mdt/*/exports/*/stats, I think it maybe not important.

**Comment by Ephemeraler** [comment]
I think the failure of the check was not caused by my code

**Comment by srebhan** [comment]
@Ephemeraler to be honest the one-line Wiki statement is not too convincing for me. By depending on an external command we force users to install additional software that might or might not be possible for them. We also do not have any control about the `lctl` command installation and it is likely to also change the output format at some point. Honestly speaking, I think the kernel interface representation in `proc` and `sys` is way more stable than a userspace tool...
To make things worst we already do have a native plugin in the code-base... I'm strongly against replacing the existing plugin giving up a zero-dependency plugin for a command-based one.

So my recommendation is to either fix the existing plugin if there is anything broken. Is there anything? If you still want to use the command, you can either make this an external plugin or write some program you can call using `inputs.exec`.

My 2ct....

**Comment by Ephemeraler** [comment]
@srebhan I am sorry for any inconvenience this may cause and appreciate your understanding. I will provide further explanation
1) The wiki is actually the lustre official statement, I use it because it give the conclusion directly. I will give some examples to explain why commands are more suitable. a) For /proc/fs/lustre/*/*/stats, /proc/fs/lustre/*/*/job_stats, the statistics function is not enabled by default. You must modify some configuration of lustre to enable it, but the command lctl can get it. This is just one example. What's more, monitoring lustre is not only part the parts provided by existing plugins, more meaningful data could't get from /proc and other filesystem. for the active state of volume, whether the ost is degraded and so on.
2) As a lustre user, the command about lctl is almost never uninstalled, it is usually used to set paramters during runtime. So I think the command is more useful and powerful to collect.
Thanks.

**Comment by powersj** [comment]
> I use it because it give the conclusion directly.

I can understand that, however as a maintainer of a project we already have a plugin that does this. Adding a second plugin that does it a different way causes us to have twice the maintenance. We need to understand the value and gain to justify adding another plugin.

> more meaningful data could't get from /proc and other filesystem

Can you please be specific about which data is not available?

You previously mentioned that some data is not available in /proc by default. However, if this requires a user to make a change, I think that is acceptable in asking.

**Comment by powersj** [comment]
@Ephemeraler while I appreciate the work, I am going to close this. @srebhan and I both have concerns around taking on a second plugin for the same thing. If there are ways we can instead update the existing plugin I would happily see that.

**Pull request closed without merging**